google/gax/__init__.py

```diff
@@ -482,3 +486,157 @@ def __next__(self):
+            self._client.cancel_operation(self._last_operation.name)
+            self._cancelled = True
+
+        return self._cancelled
+
+    def result(self, timeout=None):
+        """Enters polling loop on OperationsApi.get_operation, and once Operation.done
+           is true, then returns Operation.response if successful or throws
+           ResultError if not successful.
+
+           This method will wait up to timeout seconds. If the call hasn't
+           completed in timeout seconds, then a concurrent.futures.TimeoutError
+           will be raised. timeout can be an int or float. If timeout is not
+           specified or None, there is no limit to the wait time.
+        """
+        if not self._poll(timeout).HasField('response'):
+            raise ResultError.from_operation_error(self._last_operation.error)
+
+        return _from_any(self._result_type, self._last_operation.response)
+
+    def exception(self, timeout=None):
+        """Similar to result(), except returns the exception if any."""
+        if self._poll(timeout).HasField('error'):
+            return self._last_operation.error
+
+        return None
+
+    def cancelled(self):
+        """Return True if the call was successfully cancelled."""
+        return self._cancelled
+
+    def done(self):
+        """Issues OperationsApi.get_operation and returns value of Operation.done."""
+        return self._get_operation().done
+
+    def add_done_callback(self, done_clbk):
+        """Enters a polling loop on OperationsApi.get_operation, and once the
+           operation is done or cancelled, calls the function with this OperationFuture.
+        """
+        def _execute_clbk(self, clbk):
+            self._poll()
+            clbk(self)
+
+        threading.Thread(target=_execute_clbk, args=(self, done_clbk)).start()
+
+    def operation_name(self):
+        """Returns the value of Operation.name from the initial Operation object
+           returned from the first call. Blocks if the first call isn't done yet.
+        """
+        return self._first_operation.name
+
+    def metadata(self):
+        """Returns the value of Operation.metadata from the initial Operation object
+           returned from the first call. Blocks if the first call isn't done yet.
+        """
+        return _from_any(self._metadata_type, self._first_operation.metadata)
+
+    def last_operation_data(self):
+        """Returns the data from the last call to OperationsApi.get_operation (or if only
+           the initial API call has been made, the data from that first call). Blocks if
+           the first call isn't done yet.
+        """
+        return self._last_operation
+
+    def _get_operation(self):
+        if not self._last_operation.done:
+            self._last_operation = self._client.get_operation(
+                self._last_operation.name, self._call_options)
+
+        return self._last_operation
+
+    def _poll(self, timeout=None):
```

Comment:
Would suggest reusing (or refactoring to make reusable) the code in `api_callable.py` that @landrito linked to. I suspect you should be able to use the `_retryable` function as-is, specifying the numerical parameters as a `RetryOptions` object.